python/ray/dag/compiled_dag_node.py

```diff
@@ -495,43 +547,90 @@ def _get_node_id(self):
-                if isinstance(arg, DAGNode):
-                    arg_idx = self.dag_node_to_idx[arg]
-                    arg_channel = self.idx_to_task[arg_idx].output_channel
-                    assert arg_channel is not None
-                    resolved_args.append(arg_channel)
-                    has_at_least_one_channel_input = True
-                else:
-                    resolved_args.append(arg)
-            # TODO: Support no-input DAGs (use an empty object to signal).
-            if not has_at_least_one_channel_input:
-                raise ValueError(
-                    "Compiled DAGs require each task to take a "
-                    "ray.dag.InputNode or at least one other DAGNode as an "
-                    "input"
+        # Create executable tasks for each actor
+        for actor_id, tasks in self.actor_to_tasks.items():
+            executable_tasks = []
+            worker_fn = None
+            for task in tasks:
+                resolved_args = []
+                has_at_least_one_channel_input = False
+                for arg in task.args:
+                    if isinstance(arg, DAGNode):
+                        arg_idx = self.dag_node_to_idx[arg]
+                        arg_channel = self.idx_to_task[arg_idx].output_channel
+                        assert arg_channel is not None
+                        resolved_args.append(arg_channel)
+                        has_at_least_one_channel_input = True
+                    else:
+                        resolved_args.append(arg)
+                # TODO: Support no-input DAGs (use an empty object to signal).
+                if not has_at_least_one_channel_input:
+                    raise ValueError(
+                        "Compiled DAGs require each task to take a "
+                        "ray.dag.InputNode or at least one other DAGNode as an "
+                        "input"
+                    )
+                executable_task = ExecutableTask(
+                    task,
+                    resolved_args,
                 )
-
+                executable_tasks.append(executable_task)
+                if worker_fn is None:
+                    worker_fn = task.dag_node._get_remote_method("__ray_call__")
+            # Sort executable tasks based on their bind index, i.e., submission order
+            # so that they will be executed in that order.
+            executable_tasks.sort(key=lambda task: task.bind_index)
+
+            # Check unsupported case: binding the same actor method to the same
+            # input multiple times
+            import itertools
+
+            for _, g in itertools.groupby(executable_tasks, lambda x: x.method_name):
+                group = list(g)
+                if len(group) > 1:
+                    resolved_args = set()
+                    for executable_task in group:
+                        if (
+                            len(
+                                set(executable_task.resolved_args).intersection(
```

Comment:
I think this could raise some false positives, like `a.foo.bind(1, x)` and `a.bar.bind(1, y)` is actually okay, but they share "1" as an argument. Can you add a test for this case?

Also, to avoid all the set comparisons, you could instead check if the same actor appears twice in a node's downstream nodes.

Reply:
Good catch. Added unit test.

There is already a stricter check that an actor method can only have one actor handle as reader (probably needs extension in future). I added the check for InputNode that there should be no duplicate actor readers. The method is simple enough that I inlined it.